Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,719 @@
+# -*- coding: utf-8 -*-
+#   This work is part of the Core Imaging Library (CIL) developed by CCPi
+#   (Collaborative Computational Project in Tomographic Imaging), with
+#   substantial contributions by UKRI-STFC and University of Manchester.
+
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+
+#   http://www.apache.org/licenses/LICENSE-2.0
+
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+
+import numpy as np
+import math
+from functools import partial
+import time
+
+
+class Sampler():
+    # TODO: Work out how to make the examples testable
+    """     
+    The user is recommended to not instantiate this class directly but instead use one of the static methods that will return instances of different samplers.
+
+    Initialises a sampler that returns and then increments indices from a sequence defined by a function.
+
+
+    Custom deterministic samplers can be created by using the `from_function` static method or by subclassing this sampler class.
+
+
+
+    Parameters
+    ----------
+
+    function : Callable[[int], int]
+        A function that takes an integer iteration number and returns an integer between 0 and num_indices.
+
+    num_indices: int
+        The sampler will select from a range of indices 0 to num_indices.
+
+    sampling_type:str default is None 
+        The sampling type used. Choose from  "sequential", "staggered", "herman_meyer" and "from_function". 
```

Comment:
The default is not in the options suggested to be chosen from. What happens with the default should be documented.